### ISSUE TITLE
 error TS2774: This condition will always return true since this function is always defined.

### ISSUE THREAD

**Issue by elbandito** [apollographql/apollo-server#5367]
When running `tsc`, I keep seeing the following error:

```
 ❯ tsc                                                                                                                                                                                                                                                                                                          
node_modules/apollo-server-core/src/utils/schemaInstrumentation.ts:75:23 - error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead?

75     if (parentPath && resolveObject) {
                         ~~~~~~~~~~~~~


Found 1 error.
```

**Comment by glasser** [comment]
This is resolved on the active development branch `release-3.0` as part of upgrading TypeScript in https://github.com/apollographql/apollo-server/pull/5165
We are getting towards the end of the AS3 release process, wrapping up migration guides and docs and the like.

In the meantime, this TS version compatibility issue only happens if your TS is configured to consume the source directly from Apollo Server's packages; if it was just looking at the `.d.ts` files that we export in `dist` then it wouldn't have this problem.  I don't know how to configure tsc to prefer `.d.ts` for `node_modules` though.

**Comment by elbandito** [comment]
@glasser Thanks for tips and insight!

**Comment by elbandito** [comment]
@glasser I not directly including `apollo-server-core` - it's coming via dependency of `apollo-gateway`.
```
"@apollo/gateway@^0.23.0":
  version "0.23.2"
  resolved "https://registry.yarnpkg.com/@apollo/gateway/-/gateway-0.23.2.tgz#df6df549af2d362837dee2cafc3e3aec15901106"
  integrity sha512-WGrlAzvBfvWGnwajlUowde2VXdkH5EtqgOLnwdYCB7FVMkLT8B/Qj6ePxK+Kbwonz7lcnEvNaL4rvF2ICGudWg==
  dependencies:
    "@apollo/federation" "^0.21.2"
    "@apollo/query-planner-wasm" "^0.1.2"
    "@types/node-fetch" "2.5.4"
    apollo-graphql "^0.6.0"
    apollo-reporting-protobuf "^0.6.0"
    apollo-server-caching "^0.5.3"
    apollo-server-core "^2.19.2"
    apollo-server-env "^3.0.0"
    apollo-server-errors "^2.4.2"
    apollo-server-types "^0.6.3"
    loglevel "^1.6.1"
    make-fetch-happen "^8.0.0"
    pretty-format "^26.0.0"
```

Do you know what version of `apollo-gateway` will have the fix you mentioned above (`release-3.0`). 
Thanks.

**Comment by glasser** [comment]
`@apollo/gateway` 0.29.1 and newer should be compatible with the 3.0 release candidates (eg `3.0.0-rc.0`).
